### PR TITLE
Fix number trainable params in finite diff

### PR DESCRIPTION
**Pull request by rmoyard**
**Context:**

When using finite diff directly on a QNode with a diff method and with Jax might results in switching the argnums between the transform and its post processing.

**Description of the Change:**

Set the trainable parameters in the transforms and use then in the post processing.

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -189,6 +189,9 @@
 
 <h3>Bug fixes 🐛</h3>
 
+* Fix a bug where trainable parameters in the post-processing of finite diff were incorrect for Jax.
```
Comment (dwierichs) [inline doc/releases/changelog-dev.md:192]: Should we mention that this is only when using the transform on QNodes manually?

File: tests/gradients/finite_diff/test_finite_difference.py [hunk 2]
```diff
@@ -1065,6 +1065,45 @@ def cost_fn(x):
         )
         assert np.allclose(res, expected, atol=tol, rtol=0)
 
+    @pytest.mark.jax
+    @pytest.mark.parametrize("dev_name", ["default.qubit"])
+    def test_jax_probs(self, dev_name, tol):
+        """Tests that the output of the finite-difference transform
+        can be differentiated using JAX, yielding second derivatives."""
```
Comment (dwierichs) [inline tests/gradients/finite_diff/test_finite_difference.py:1072]: I think this needs to be updated?

**Review by dwierichs — APPROVED** [review]
Nice, thanks for the quick fix @rmoyard!
Just to make sure I understand the bug: The `tape` was being referenced in the postprocessing function, but the `trainable_params` of the tape could change in the meantime, making the postprocessing do wrong things. Is this correct?
Are there other places in the `gradients` module that could suffer from this sharp bit? :thinking: 

Otherwise just a tiny comment on the changelog entry and a test docstring.

**Comment by rmoyard** [comment]
When setting diff_method="finite_diff" it chooses the execution with the Jax interface. In the Jax interface we set the trainable parameters, as None of them are considered trainable because there are not tracers `isinstance(tensor, jax.core.Tracer)`. Indeed nothing is jacobian or grad is not used. Users should not choose a diff method when applying a transform but that's possible.

There is a stability issue with Torch, the execution pipeline default vs torch have different results.

**Comment by trbromley** [comment]
> Also, does this PR have anything to do with Issue https://github.com/PennyLaneAI/pennylane/issues/4882 ?

They're related with similar example code (except the interface). In #4882, Torch gives incorrect gradients when `diff_method` is specified. Here, the issue is that JAX raises an error when `diff_method` is _not_ specified. :confused:

TLDR I don't think this PR helps with #4882.

**Review by timmysilv — APPROVED** [review]
